controllers/event/eventsource_controller.go

```diff
@@ -0,0 +1,344 @@
+// move the current state of the cluster closer to the desired state.
+// TODO(user): Modify the Reconcile function to compare the state specified by
+// the EventSource object against the actual cluster state, and then
+// perform operations to make the cluster state reflect the state specified by
+// the user.
+//
+// For more details, check Reconcile and its Result here:
+// - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.8.3/pkg/reconcile
+func (r *EventSourceReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	r.ctx = ctx
+	log := r.Log.WithValues("EventSource", req.NamespacedName)
+	log.Info("eventsource reconcile starting...")
+
+	var es openfunctionevent.EventSource
+	r.envs = &SourceEnvConfig{}
+
+	if err := r.Get(ctx, req.NamespacedName, &es); err != nil {
+		log.V(1).Info("EventSource deleted", "error", err)
+		return ctrl.Result{}, util.IgnoreNotFound(err)
+	}
+
+	if _, err := r.createOrUpdateEventSource(&es); err != nil {
+		log.Error(err, "Failed")
+		return ctrl.Result{}, err
+	}
+
+	return ctrl.Result{}, nil
+}
+
+// 1. Retrieve the configuration of the event bus based on the value of spec.eventBusNames (Used to connect event sources to event bus)
+// 2. Generate dapr component specs (source and sink) for the event source
+// 3. Create dapr component by using specs above
+// 4. Generate an event source configuration SourceEnvConfig
+// 5. Create event source handler workload (pass SourceEnvConfig in)
+func (r *EventSourceReconciler) createOrUpdateEventSource(es *openfunctionevent.EventSource) (ctrl.Result, error) {
+	log := r.Log.WithName("createOrUpdate")
+
+	// Retrieve the configuration of the event bus based on the value of spec.eventBusNames
+	eventBusComponentConfigs, err := r.retrieveEventBusConfigs(es)
+	if err != nil {
+		log.Error(err, "Failed to retrieve event bus component names", "namespace", es.Namespace, "name", es.Name)
+		return ctrl.Result{}, err
+	}
+
+	// Converts the event source specification into a connector
+	// containing the specification for creating the dapr component.
+	connectors, err := adapters.NewEventSourceConnectors(es)
+	if err != nil {
+		log.Error(err, "Failed to generate connectors")
+		return ctrl.Result{}, err
+	}
+
+	// Create dapr component to adapter the event source sink
+	sinkComponentName, err := r.createOrUpdateSink(es)
+	if err != nil {
+		log.Error(err, "Failed to create sink", "namespace", es.Namespace, "name", es.Name)
+	}
+
+	// At least a EventBus exists, or spec.sink needs to be configured in the EventSource for the EventSource to work properly
+	// If spec.eventBusName is not set, then the EventBus named "default" will be checked to see if it exists
+	if sinkComponentName == "" && eventBusComponentConfigs == nil {
+		log.Error(err, "Failed to find eventbus nor sink", "namespace", es.Namespace, "name", es.Name)
+		return ctrl.Result{}, err
+	}
+
+	// Create a one-to-one workload for each connector
+	for _, c := range connectors {
```

Comment:
We'd better to use hash to check whether the crd has been changed and needs to be updated

https://github.com/OpenFunction/OpenFunction/blob/main/controllers/function_controller.go#L117

Same comments to all below createOrUpdatexxx